<!DOCTYPE html>
<html>
<body>

<center>
<h1>Tufts University DS288 - Data Science Master's Project</h1>
  <h2> Edward Moseley </h2>
</center>

<br>
<br>

<h1>Project Notebook</h1>

<h3>(Preliminary) Project Description</h3>

<p> This Master's Capstone is an investigation into various Bioinformatics/Computational Biology tools used for the prediction of protein-protein/protein-small molecule interactions, and/or the prediction of Gene Ontologies (GO) on the basis of protein sequence data, genome wide association studies (GWAS), or other data sources.</p>

<p>The project will begin fairly broad in scope, but is expected to narrow significantly through meetings with professors and possible collaborators. If a project with privately-generated data is not available, public data sources will be employed.</p>

<br>
  
  <h2> Week of Monday, May 22nd </h2>
  
  <p>Reached out to individuals in the Bioinformatics and Computional Biology (BCB) working group expressing interest in studying: </p>
  
  <ul>
    <li>Protein Gene Ontology (GO) Labeling (Prof. LC)</li>
    <li>Inferring Protein-Protein Interactions (Prof. LC)</li>
    <li>Precision Medicine (Prof. DS) </li>
  </ul>
  
  <p>Met with Prof. LC, discussed research interests and was provided with two publications to read:</p>
    
    <ul>
      <li>Singh, R., et al. "Topsy-Turvy: integrating a global view into sequence-based PPI prediction", <i>Bioinformatics</i> (2022)</li>
      <li>Devkota, K., et al. "GLIDER: function prediction from GLIDE-based neighborhoods", <i>Bioinformatics</i> (2022) </li>
    </ul>
  
<br>

<h2> Week of Monday, May 29th</h2>
  
<p>Read publications from previous week, and began to plan meetings with professors from the School of Veterinary Medicine relating to a project on antiody heavy chains and virulence factors associated with Luke Worms (perasites which cause fairly grotesque illnesses animals).</p>

  <p>New publication (not yet in press) added to the must-read list:</p>

  <ul>
        <li>Singh, R., et al. "Learning the Drug-Target Interaction Lexicon", <i>bioRxiv</i> (2022)</li>
  </ul>

<br>

<h2> Week of Monday, June 5th</h2>

<p>Read publication from previous week, preparing for first meeting of the BCG Working Group (with Professors DS and LC) on Friday, June 9th at 10am, for introductions and to discuss more project ideas/details.</p>

<p>This is also the first week meeting with Prof. MA, to help flesh out the project and discuss expectations moving forward.</p>

<br>


<h2> Week of Monday, Jun 12th</h2>
  
<p>Attended BCB working group meeting and spent time with students/professors to discuss interesting research tools and ideas afterwards.</p>

<p>On Tuesday, June 13th, I met with a number of individuals from the School of Veterinary Medicine, as well as external collaborators from MIT and industry.</p>

The general project relates to the study of "VHHs", which are the variable domains of heavy-chain antivodies (HCAb), which are specific to the Camelid biological family. The overview of the biological experimental design was reviewed in detail, and includes:

<h3>(Animal) Study Design</h3>

<ol type="1">
  <li>Pre-Immunization blood-draw, Peripheral Blood Lymphocyte (PBL) isolation, DNA sequencing to determine Pre-Immunization VHH repertoir.</li>
  <li>Prime/Boost regimen, where blood draws are performed immediately prior to immunization and 2-weeks post-vaccination for PBL isolation and VHH sequencing.</li>
</ol>

<h3>Bioinformatics Study Design</h3>

<p>For the bioinformatics analysis, we still have some details to iron out. However, the overall objective appears to relate to tracking the evolution of these cells through the process of somatic hypermutation and clonal selection.</p>

<p>It is unclear, at this point, if we will attempt to model the interaction between the antibodies and the antigenic material the animals were immunized against. However, it would be very interesting to see if we could use the VHH amino acid sequence data to predict the avidity (per protein-capture ELISA titers, or SPR data via Biacore) with which the VHHs would bind to the antigenic material they received during immunization.</p>

<br>

<h2> Week of Monday, Jun 19th</h2>

<p>Access to research data repository granted, and data were downloaded.</p> 

<p>This week, the focus is heavily on the "Immcantation" Python/R package (https://immcantation.readthedocs.io/en/stable/), which is useful for analyzing the germline and somatic diversity of the adaptive immune receptor repertoire (AIRR).</p>

<br>

<h2> Week of Monday, Jun 26th</h2>

<p>This week, the focus remained on the "Immcantation" Python/R library, with a specific focus on pRESTO (The REpertoire Sequencing TOolkit).</p>

<p>Began prototyping pipeline for:</p>

<ol type="1">
  <li>Antibody Paired-End Assembly.</li>
  <li>Quality Control/Primer Annotation.</li>
  <li>Deduplication and Filtering.</li>
</ol>

<br>

<h2> Week of Monday, Jul 3rd</h2>

<p>This week, the initial pipeline was completed using dummy data (although "real" data are available). At each step in the pipeline, logging data was printed to files and stored.</p>

<p>Focus has also transitioned slightly to a back-up plan in the event that there are insufficient data from the ongoing Alpaca vaccine study to statisfy course requirements. As a result, we consider:</p>

<ol type="1">
  <li>Using PPI model to predict binding of antibody sequences to HIV-1 Env (relying on the Los Alamos National Laboratory database).</li>
  <li>Possible to discriminate between neutralizing/non-neutralizing binding antibodies?</li>
  <li>Does the LANL database have meaningful germline data?</li>
  <li>Can these models generalize to other viruses, like SARS-CoV-2 (LANL appears to have data)</li>
</ol>

<br>
<br>
  
</body>
</html>
